be/src/vec/aggregate_functions/aggregate_function_foreach.h

```diff
@@ -0,0 +1,263 @@
+                ensure_aggregate_data(place, rhs_state.dynamic_array_size, *arena);
+
+        const char* rhs_nested_state = rhs_state.array_of_aggregate_datas;
+        char* nested_state = state.array_of_aggregate_datas;
+
+        for (size_t i = 0; i < state.dynamic_array_size && i < rhs_state.dynamic_array_size; ++i) {
+            nested_function->merge(nested_state, rhs_nested_state, arena);
+
+            rhs_nested_state += nested_size_of_data;
+            nested_state += nested_size_of_data;
+        }
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, BufferWritable& buf) const override {
+        const AggregateFunctionForEachData& state = data(place);
+        write_binary(state.dynamic_array_size, buf);
+        const char* nested_state = state.array_of_aggregate_datas;
+        for (size_t i = 0; i < state.dynamic_array_size; ++i) {
+            nested_function->serialize(nested_state, buf);
+            nested_state += nested_size_of_data;
+        }
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, BufferReadable& buf,
+                     Arena* arena) const override {
+        AggregateFunctionForEachData& state = data(place);
+
+        size_t new_size = 0;
+        read_binary(new_size, buf);
+
+        ensure_aggregate_data(place, new_size, *arena);
+
+        char* nested_state = state.array_of_aggregate_datas;
+        for (size_t i = 0; i < new_size; ++i) {
+            nested_function->deserialize(nested_state, buf, arena);
+            nested_state += nested_size_of_data;
+        }
+    }
+
+    void insert_result_into(ConstAggregateDataPtr __restrict place, IColumn& to) const override {
+        const AggregateFunctionForEachData& state = data(place);
+
+        auto& arr_to = assert_cast<ColumnArray&>(to);
+        auto& offsets_to = arr_to.get_offsets();
+        IColumn& elems_to = arr_to.get_data();
+
+        char* nested_state = state.array_of_aggregate_datas;
+        for (size_t i = 0; i < state.dynamic_array_size; ++i) {
+            nested_function->insert_result_into(nested_state, elems_to);
+            nested_state += nested_size_of_data;
+        }
+
+        offsets_to.push_back(offsets_to.back() + state.dynamic_array_size);
+    }
+
+    bool allocates_memory_in_arena() const override {
+        return nested_function->allocates_memory_in_arena();
+    }
+
+    bool is_state() const override { return nested_function->is_state(); }
+
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena* arena) const override {
+        const IColumn* nested[num_arguments];
+
+        for (size_t i = 0; i < num_arguments; ++i) {
+            nested[i] = &assert_cast<const ColumnArray&>(*columns[i]).get_data();
+        }
+
+        const auto& first_array_column = assert_cast<const ColumnArray&>(*columns[0]);
+        const auto& offsets = first_array_column.get_offsets();
+
+        size_t begin = offsets[row_num - 1];
```

Comment:
what happen if row_num == 0? maybe overflow access mem offset

Reply:
The index of offsets can be -1.
// [[2,1,5,9,1], [1,2,4]] --> data column [2,1,5,9,1,1,2,4], offset[-1] = 0, offset[0] = 5, offset[1] = 8

Reply:
Warning ！！！row_num is `size_t` -1 = 18446744073709551615

Reply:
my mistake.